keps/1397-evictions-in-background/README.md

```diff
@@ -0,0 +1,1122 @@
+  `MaxNoOfPodsEvacuationsPerNamespace` feature gated v1alpha1 fields.
+- **New configurable list of eviction message prefixes (introduced later when needed)**:
+  each third party validating admission webhook can provide different response message.
+  It is not feasible to ask every party to change their messages for a temporary solution
+  until the evacuation API is in place. Thus, temporary introducing a new feature
+  gated `evictionMessagePrefixes` field (`[]string`) for a list of known prefixes
+  signaling an eviction in background.
+  Normally, "Eviction triggered" prefix will be acknowledged by default.
+
+#### v1alpha1
+
+- Keep track of evictions in progress: A new cache will be introduced
+  to keep track of pods annotated with `descheduler.alpha.kubernetes.io/request-evict-only`
+  that were requested to be evicted through the upstream eviction API.
+  In case a descheduler gets unexpectedly restarted any pod annotated
+  with `descheduler.alpha.kubernetes.io/eviction-in-progress` will be added
+  to the cache if not already present.
+  The cache will help with sorting and avoiding a duplicated eviction
+  request of pods that were not annotated
+  with `descheduler.alpha.kubernetes.io/eviction-in-progress` quickly enough.
+  Each item in the cache will have a TTL attribute to address cases where an eviction
+  request was intercepted by an external component but the component failed to annotate the pod.
+- Implement a new built-in sorting plugin: Extend the descheduling framework
+  with a new sorting plugin that will prefer pods
+  with `descheduler.alpha.kubernetes.io/eviction-in-progress` annotation or those
+  that are already in the cache of to-be-evicted pods.
+  The plugin will be enabled when the new functionality is enabled (through a feature gate).
+  Each plugin will have option to either take the new sorting into account or not.
+  The sorting plugin can be executed either in pre-filter or pre-eviction phase.
+- Implement a multi-level sorting: Allow plugins to first sort pods based
+  on eviction-in-progress quality and then based on configured sorting plugins.
+  Preferring pods that are already in a process of eviction over other pods
+  is a crucial step for reducing unnecessary disruptions.
+- Internal counters aware of eviction in progress: At the beginning of each
+  descheduling loop all internal counters will take into account pods
+  that are either annotated with `descheduler.alpha.kubernetes.io/eviction-in-progress`
+  or are present in the cache of to-be-evicted pods.
+
+#### v1alphaN or beta
+- Replace the current eviction API with the new
+  [evacuation API](https://github.com/kubernetes/enhancements/pull/4565) by default.
+  Both `descheduler.alpha.kubernetes.io/request-evict-only` and
+  `descheduler.alpha.kubernetes.io/eviction-in-progress` will be deprecated.
+  Existing deployments will need to remove `descheduler.alpha.kubernetes.io/request-evict-only`
```

Comment:
does that mean that delayed eviction/evacuation will not work for the workloads that forget to remove the annotation?

Reply:
Yes, this can happen until the deprecated annotations get removed (2 releases minimum after the deprecation is announced). This could be a blocking reason for a broader adoption.

Reply:
Done